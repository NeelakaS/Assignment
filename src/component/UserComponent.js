import React, {useEffect} from 'react';
import {
  View,
  Text,
  FlatList,
  TouchableOpacity,
  ActivityIndicator,
  Image,
} from 'react-native';
import {connect} from 'react-redux';
import {fetchUserData} from '../action/actions';
import styles from './UserComponent.style';

type UserData = {
  id: number,
  name: string,
  username: string,
  email: string,
  phone: string,
  website: string,
};

type Props = {
  fetchUserInfo: () => void,
  userData: UserData,
  navigation: Objct,
  loading: boolean,
};

const UserComponent = (props: Props) => {
  const {fetchUserInfo, userData, navigation, loading} = props;

  // hooks
  useEffect(() => {
    // getting user details
    fetchUserInfo;
  });

  const extractKey = (item: UserData, index: number) => `${item?.id}-${index}`;

  const renderItem = ({item}: UserData) => {
    return (
      <TouchableOpacity
        style={styles.listContainer}
        onPress={() => {
          navigation.navigate('Album', {
            userId: item.id,
          });
        }}>
        <View style={styles.userImage}>
          <Image source={require('./img/user_thumbnail.png')} />
        </View>
        <View>
          <Text style={styles.userName}>{item.name}</Text>
          <Text>Email: {item.email}</Text>
          <Text>Phone: {item.phone}</Text>
          <Text>Website: {item.website}</Text>
        </View>
      </TouchableOpacity>
    );
  };

  // loading untill API receives data
  if (loading) {
    return (
      <View style={styles.actionIndicator}>
        <ActivityIndicator />
      </View>
    );
  }

  return (
    <FlatList
      ItemSeparatorComponent={() => <View style={styles.listItemSeparator} />}
      data={userData}
      renderItem={renderItem}
      keyExtractor={extractKey}
    />
  );
};

const mapDispatchToProps = dispatch => ({
  fetchUserInfo: dispatch(fetchUserData()),
});
const mapStateToProps = state => ({
  userData: state.userData.data,
  loading: state.userData.isFetching,
});

// eslint-disable-next-line prettier/prettier
export default connect(mapStateToProps, mapDispatchToProps)(UserComponent);
