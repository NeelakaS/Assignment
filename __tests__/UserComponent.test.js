import 'react-native';
import React from 'react';

import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import saga from 'redux-saga';

import UserComponent from '../src/component/UserComponent';

const USER_DATA = [
  {
    id: 1,
    name: 'Leanne Graham',
    username: 'Bret',
    email: 'Sincere@april.biz',
    phone: '1-770-736-8031 x56442',
    website: 'hildegard.org',
  },
  {
    id: 2,
    name: 'Ervin Howell',
    username: 'Antonette',
    email: 'Shanna@melissa.tv',
    phone: '010-692-6593 x09125',
    website: 'anastasia.net',
  },
];

const DEFAULT_PROPS = {
  fetchUserData: jest.fn(),
  userData: [],
  navigation: {
    getParam: jest.fn(),
  },
  loading: false,
};

const middlewares = [saga];
const mockStore = configureMockStore(middlewares);
let wrapper;
describe('User Component', () => {
  it('render UserComponent correctly', () => {
    const initialState = {
      userData: {
        data: USER_DATA,
        isDataFetched: false,
        isFetching: false,
        isError: false,
      },
    };
    const store = mockStore(initialState);

    wrapper = renderer
      .create(
        <Provider store={store}>
          <UserComponent {...DEFAULT_PROPS} />
        </Provider>,
      )
      .toJSON();
    expect(wrapper).toMatchSnapshot();
  });
});
