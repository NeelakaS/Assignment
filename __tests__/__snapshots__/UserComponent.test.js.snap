// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Component render UserComponent correctly 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  data={
    Array [
      Object {
        "email": "Sincere@april.biz",
        "id": 1,
        "name": "Leanne Graham",
        "phone": "1-770-736-8031 x56442",
        "username": "Bret",
        "website": "hildegard.org",
      },
      Object {
        "email": "Shanna@melissa.tv",
        "id": 2,
        "name": "Ervin Howell",
        "phone": "010-692-6593 x09125",
        "username": "Antonette",
        "website": "anastasia.net",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "flexDirection": "row",
            "height": 100,
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "marginHorizontal": 10,
              "width": "15%",
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/component/img/user_thumbnail.png",
              }
            }
          />
        </View>
        <View>
          <Text
            style={
              Object {
                "fontSize": 17,
                "fontWeight": "bold",
                "lineHeight": 25,
              }
            }
          >
            Leanne Graham
          </Text>
          <Text>
            Email: 
            Sincere@april.biz
          </Text>
          <Text>
            Phone: 
            1-770-736-8031 x56442
          </Text>
          <Text>
            Website: 
            hildegard.org
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#404041",
            "minHeight": 0.5,
            "opacity": 0.3,
          }
        }
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "flexDirection": "row",
            "height": 100,
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "marginHorizontal": 10,
              "width": "15%",
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/component/img/user_thumbnail.png",
              }
            }
          />
        </View>
        <View>
          <Text
            style={
              Object {
                "fontSize": 17,
                "fontWeight": "bold",
                "lineHeight": 25,
              }
            }
          >
            Ervin Howell
          </Text>
          <Text>
            Email: 
            Shanna@melissa.tv
          </Text>
          <Text>
            Phone: 
            010-692-6593 x09125
          </Text>
          <Text>
            Website: 
            anastasia.net
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
